nomad/structs/devices.go

```diff
@@ -0,0 +1,130 @@
+package structs
+
+// DeviceAccounter is used to account for device usage on a node. It can detect
+// when a node is oversubscribed and can be used for deciding what devices are
+// free
+type DeviceAccounter struct {
+	Devices map[DeviceIdTuple]*DeviceAccounterInstance
+}
+
+// DeviceAccounterInstance wraps a device and adds tracking to the instances of
+// the device to determine if they are free or not.
+type DeviceAccounterInstance struct {
+	// Device is the device being wrapped
+	Device *NodeDeviceResource
+
+	// Instances is a mapping of the device IDs of the instances to their usage.
+	// Only a value of 0 indicates that the instance is unused.
+	Instances map[string]int
+}
+
+// NewDeviceAccounter returns a new device accounter. The node is used to
+// populate the set of available devices based on what healthy device instances
+// exist on the node.
+func NewDeviceAccounter(n *Node) *DeviceAccounter {
+	numDevices := 0
+	var devices []*NodeDeviceResource
+
+	// COMPAT(0.11): Remove in 0.11
```

Comment:
since devices are only going to be in NodeResources, isn;t this comment inaccurate?

Reply:
Post 0.11 the if nil check can be removed since all nodes will have that field set